### PR TITLE
Refactor note toggle (complete)

### PR DESCRIPTION
**Pull request by cutandpastey**
This PR completes the `note-toggle` refactoring.

### QA
- Test un-note part of a note. Make a selection within a note and hit F10
- Test un-note a whole note. Drop the caret within a note and hit F10

### CODE REVIEW

File: src/actions/noting/create-note-from-selection.js [hunk 1]
```diff
@@ -0,0 +1,94 @@
...
+var isParagraph = require('../../utils/vfocus/is-paragraph');
+var createVirtualScribeMarker = require('../../utils/create-virtual-scribe-marker');
+var isNotEmpty = require('../../utils/vfocus/is-not-empty');
+var removeEmptyNotes = require('../../actions/noting/remove-empty-notes');
+
+// treeFocus: tree focus of tree containing two scribe markers
+// Note that we will mutate the tree.
+module.exports = function createNoteFromSelection(focus){
+
+  if (!isVFocus(focus)){
```
Comment (hmgibson23) [inline src/actions/noting/create-note-from-selection.js]: Style :)

File: src/actions/noting/create-note-from-selection.js [hunk 2]
```diff
@@ -0,0 +1,94 @@
...
+  // Replace the nodes in the tree with the wrapped versions.
+  _.zip(toWrapAndReplace, wrappedTextNodes).forEach(focusAndReplacementVNode => {
+    var focus = focusAndReplacementVNode[0];
+    var replacementVNode = focusAndReplacementVNode[1];
+    focus.replace(replacementVNode);
+  });
+
+  // If we end up with an empty note a <BR> tag would be created. We have to do
+  // this before we remove the markers.
+  removeErroniousBrTags(focus);
```
Comment (hmgibson23) [inline src/actions/noting/create-note-from-selection.js]: This is erroneous...

File: src/actions/noting/create-note-from-selection.js (hunk 1 above)
Comment (cutandpastey) [inline src/actions/noting/create-note-from-selection.js]: As mentioned previously theres a `.jscsrc` on the way which will fix these if you bear it in the code base for now?

File: src/actions/noting/create-note-from-selection.js [hunk 3]
```diff
@@ -0,0 +1,94 @@
...
+
+  // We want to place the caret after the note. First we have to remove the
+  // existing markers.
+  removeScribeMarkers(focus);
+
+  // Update note properties (merges if necessary).
+  var lastNoteSegment = findLastNoteSegment(toWrapAndReplace[0]);
+  var noteSegments = findEntireNote(lastNoteSegment);
+  resetnoteSegmentClasses(noteSegments);
```
Comment (hmgibson23) [inline src/actions/noting/create-note-from-selection.js]: Can this be camel case to be consistent?

File: src/actions/noting/create-note-from-selection.js [hunk 4]
```diff
@@ -0,0 +1,94 @@
...
+  // TODO: Revisit our caching strategy to make it less of a "foot gun", or
+  // refactor so that we do less tree traversals and remove the caching.
+  notesCache.set(focus);
+
+  // Now let's place that caret.
+  var outsideNoteFocus = noteSegments.splice(-1)[0].find(isNotWithinNote);
+
+  // We guard against the case when the user notes the last piece of text in a
+  // Scribe instance. In that case we don't bother placing the cursor.
+  // (What behaviour would a user expect?)
```
Comment (hmgibson23) [inline src/actions/noting/create-note-from-selection.js]: It's worth turning this questions into an issue or else we'll just forget that we have some behaviour that should probably be clarified at some point.

File: src/actions/noting/remove-erroneous-br-tags.js [hunk 5]
```diff
@@ -0,0 +1,71 @@
...
+  var segments = [
+    marker.find(isNoteSegment, 'prev'),
+    marker.find(isNoteSegment)
+  ].filter(o => !!o);
+
+  // Replace/delete empty notes, and parents that might have become empty.
+  segments.map(segment => {
+    if (hasOnlyEmptyTexChildren(segment)){
+      // When we delete a space we want to add a space to the previous
+      // note segment.
```
Comment (hmgibson23) [inline src/actions/noting/remove-erroneous-br-tags.js]: I don't really understand why we would delete a space?

File: src/actions/noting/remove-part-of-note.js [hunk 6]
```diff
@@ -0,0 +1,94 @@
...
+
+  if (!isVFocus(focus)){
+    errorHandle('Only a valid VFocus can be passed to unNotePartOfNote, you passed: %s', focus);
+  }
+
+  var focusesToUnnote = findTextBetweenScribeMarkers(focus);
+  var entireNote = findEntireNote(focusesToUnnote[0]);
+
+  var entireNoteTextNodeFocuses = _(entireNote).map(flattenTree)
+  .flatten().value().filter(isVText);
```
Comment (hmgibson23) [inline src/actions/noting/remove-part-of-note.js]: Indentation but I'm assuming js-whatever will fix it.

Comment (cutandpastey) [inline src/actions/noting/remove-part-of-note.js]: hopefully, @robinedman @AleLudovici ?

File: src/actions/noting/remove-part-of-note.js [hunk 7]
```diff
@@ -0,0 +1,94 @@
...
+
+  var entireNoteTextNodes = entireNoteTextNodeFocuses.map(nodeFocus => nodeFocus.vNode);
+  var textNodesToUnote = focusesToUnnote.map(nodeFocus => nodeFocus.vNode);
```
Comment (hmgibson23) [inline src/actions/noting/remove-part-of-note.js]: Think this should be ```textsNodetoUnnote```

**Comment by hmgibson23** [comment]
:+1:

**Pull request merged**